include/revng/Pipes/PipelineManager.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <vector>
+
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Runner.h"
+#include "revng/Pipeline/SavableObject.h"
+#include "revng/Pipes/LLVMContextWrapper.h"
+#include "revng/Pipes/ModelGlobal.h"
+
+namespace revng::pipes {
+
+/// This is a "God Object" that can be used when there will be exactly one
+/// Pipeline spawned by this process.
+///
+/// It will care of all configurations including llvm ones and provide a runner
+/// in a usable state.
+class PipelineManager {
+private:
+  explicit PipelineManager() = default;
+
+  std::string ExecutionDirectory;
+  std::unique_ptr<LLVMContextWrapper> Context;
+  std::unique_ptr<ModelGlobal> ModelWrapper;
+  std::unique_ptr<pipeline::Context> PipelineContext;
+  std::unique_ptr<pipeline::Loader> Loader;
+  std::unique_ptr<pipeline::Runner> Pipeline;
```

Comment:
this is done to keep pointers stable most of all, and to delay the creation of the object since they require complex initialization. It is not a issue because this class is just  a thin wrapper that all it does is to keep everything used by the c api in a single place.

Reply:
> this is done to keep pointers stable most of all

Is this FUD or is it actually necessary?

> to delay the creation of the object since they require complex initialization

If you're using `std::unique_ptr` for delayed initialization purposes, you could use `std::optional` as well, but please try to minimize this (and maybe even reconsider the various things to have a default constructor, if they don't have it).

Reply:
i can't do this simply. The construction of this object are failable, so there is a constructor method
```
static llvm::Expected<PipelineManager> Pipeline::create(llvm::ArrayRef<std::string> PipelinePath, llvm::ArrayRef<std::string> EnablingFlags, llvm::StringRef ExecutionDirectory);
```
since expected is allocated on the stack, the pointers contained would be moved and thus references would break. I can return a Expected<unique_ptr> to avoid the issue, but the current solution is a bit more complex internally to be less complex outside.

Reply:
that code snippet is still UB i think. 
rvo does not guarantee that the move will not happen, i think. not sure toh